drivers/ethernet/phy/phy_microchip_ksz8081.c

```diff
@@ -0,0 +1,459 @@
+	if (ret) {
+		LOG_ERR("Error writing phy (%d) advertising register", config->addr);
+		k_mutex_unlock(&data->mutex);
+		return ret;
+	}
+
+	/* (re)do autonegotiation */
+	ret = phy_mc_ksz8081_autonegotiate(dev);
+	if (ret) {
+		LOG_ERR("Error in autonegotiation");
+		return ret;
+	}
+
+	/* Unlock mutex */
+	k_mutex_unlock(&data->mutex);
+
+
+	/* Get link status */
+	ret = phy_mc_ksz8081_get_link(dev, &data->state);
+	if (ret) {
+		return ret;
+	}
+
+	/* Start monitoring */
+	k_work_schedule(&data->phy_monitor_work,
+				K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+
+	/* Log the results of the configuration */
+	LOG_INF("PHY %d is %s", config->addr, data->state.is_up ? "up" : "down");
+	LOG_INF("PHY (%d) Link speed %s Mb, %s duplex\n", config->addr,
+		(PHY_LINK_IS_SPEED_100M(data->state.speed) ? "100" : "10"),
+		PHY_LINK_IS_FULL_DUPLEX(data->state.speed) ? "full" : "half");
+
+	return 0;
+}
+
+static int phy_mc_ksz8081_link_cb_set(const struct device *dev,
+					phy_callback_t cb, void *user_data)
+{
+	struct mc_ksz8081_data *data = dev->data;
+
+	data->cb = cb;
+	data->cb_data = user_data;
+
+	phy_mc_ksz8081_get_link(dev, &data->state);
+
+	data->cb(dev, &data->state, data->cb_data);
+
+	return 0;
+}
+
+static void phy_mc_ksz8081_monitor_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct mc_ksz8081_data *data =
+		CONTAINER_OF(dwork, struct mc_ksz8081_data, phy_monitor_work);
+	const struct device *dev = data->dev;
+	struct phy_link_state state;
+	int rc;
+
+	rc = phy_mc_ksz8081_get_link(dev, &state);
+
+	if (rc == 0 && memcmp(&state, &data->state, sizeof(struct phy_link_state))) {
+		memcpy(&data->state, &state, sizeof(struct phy_link_state));
+		if (data->cb) {
+			data->cb(dev, &data->state, data->cb_data);
+		}
+	}
+
+	/* TODO change this to GPIO interrupt driven */
+	k_work_reschedule(&data->phy_monitor_work,
+				K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+}
+
+static int phy_mc_ksz8081_init(const struct device *dev)
+{
+	const struct mc_ksz8081_config *config = dev->config;
+	struct mc_ksz8081_data *data = dev->data;
+	uint32_t bmcr;
+	int ret;
+
+	data->dev = dev;
+
+	ret = k_mutex_init(&data->mutex);
+	if (ret) {
+		return ret;
+	}
+
+	mdio_bus_enable(config->mdio_dev);
+
+	/* Prevent NAND TREE mode */
+	ret = gpio_pin_configure_dt(&config->interrupt_gpio, GPIO_OUTPUT_ACTIVE);
+	if (ret) {
+		return ret;
+	}
+
+	/* Start reset */
+	ret = gpio_pin_configure_dt(&config->reset_gpio, GPIO_OUTPUT_INACTIVE);
+	if (ret) {
+		return ret;
+	}
+
+	/* Wait for 500 ms as specified by datasheet */
+	k_busy_wait(USEC_PER_MSEC * 500);
+
+	/* Reset over */
+	ret = gpio_pin_set_dt(&config->reset_gpio, 1);
+	if (ret) {
+		return ret;
+	}
+
+	ret = phy_mc_ksz8081_read(dev, MII_BMCR, &bmcr);
+	if (ret) {
+		LOG_ERR("Error reading phy (%d) basic control register", config->addr);
+		return ret;
+	}
+
+	/* Software MDIO reset */
+	bmcr |= MII_BMCR_RESET;
+
+	ret = phy_mc_ksz8081_write(dev, MII_BMCR, bmcr);
```

Comment:
Is it necessary to perform a software reset right after a hardware reset? Since `drivers/ethernet/eth_nxp_enet.c:571` performs a busy wait after a software reset, it should be done here as well if the software reset is still required ofcourse.

Reply:
I think this was left in by accident, I will remove it, it is not necessary, thanks